### ISSUE TITLE
New version in TikZ?

### ISSUE THREAD

**Issue by i2000s** [CQuIC/qcircuit#6]
Anyone plan to do it or not? Seems not not that hard though...

**Comment by i2000s** [comment]
A followup idea on how to implement this package in TikZ (inspired by a [Stackexchange thread](http://tex.stackexchange.com/questions/223069/qcircuit-how-to-insert-a-new-line-in-a-gate)):

* All circuit lines can be drawn as straight lines in the background at first. 

* The overall layout is defined by a matrix environment. We need to write a wrapper to calculation the number of elements of inputs and the coordinates of each elements. 

* All operators are drawn as geometric shapes filled with the same color as the paper has, and should be put on top of the background circuits. Of most cases, operators are just boxes, and should be able to defined as a style element. There shouldn't be overlaps between different operators. For instance, we can define **X**-gates and a customizable gate by 
```
\tikzset{
X-box/.style  = {draw,fill=white,minimum size= 0.5 cm, text width= a cm, align=center},
operator/.style = {draw,fill=white,minimum height=7em,text width=b cm, align=center},
...
}
...
\newcommand{\x-gate}[1]{\node[X-box] (place_a) {X} };
\newcommand{\operator}[1]{\node[operator] (place_b) {#1} };
\newcommand{\qcircuit}[1]{ 
    \matrix[row sep=0.4cm, column sep=0.8cm] (circuit) { #1}
    \begin{pgfonlayer}{background}
        \draw[thick] (Starting_point_with_decorations) -- (Box_1.west)  (Point_1) -- (Box_M.west); % Draw the background line.
       ...
    \end{gpfonlayer} };
```
When a user want to draw a q-circuit, he will just use
```
\qcircuit{ \x-gate & \operator{$\ket{\psi}$} \\ 
              \y-gate &  & }
```
basically the same UI style as the original package defines.

**Comment by Travis-S** [comment]
One thing I worry about though is the UX (user experience) - Jonathan had
observed the current style is incredibly user friendly.

This being said, I like the idea of a wrapper.

~Travis Scholten

**Comment by sflammia** [comment]
I would suggest that any rewrite in TikZ should have a different package name. I recommend "qcircuitz", with the "z" at the end to signify TikZ. Also, the syntax should allow for specifying keys in the natural way, e.g.,
\qcircuit[measurement=meter, gateheight=12pt]{ <stuff> }

**Comment by sflammia** [comment]
You don't want to use a white background to "paste over" the wires in the background. It can cause incompatibilities if someone wants the circuit background to match the color behind the gates (think of a powerpoint presentation, for example). It is better to actually have wires end at gates and let the gate background color be a variable.

**Comment by i2000s** [comment]
A tentative prototype has been implemented in [a new repo--qcircuitz](https://github.com/CQuIC/qcircuitz). Maybe more discussions about the design and technical details can go there.
